Feature/gat

youph opened the pull request:
This PR addresses #216 and includes the following:
- [x] Graph Attention layer implementation
- [x] GAT class implementation (a stack of Graph Attention layers)
- [x] FullBatchNodeGenerator class
- [x] Demos:
  - [x] python script
  - [x] notebook
- [x] Unit tests
- [x] README updated
  - [x] Reference to GAT paper added
  - [x] GAT algorithm added
- [x] AUTHORS updated with GAT authors

PantelisElinas commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+
+        assert isinstance(x_inp, list), "input must be a list, got {} instead".format(
+            type(x_inp)
+        )
```
See my comment above about assert vs raising ValueError.

youph replied:
Fixed

PantelisElinas commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -448,3 +454,132 @@ def flow_from_dataframe(self, node_targets):
+        self.node_list = list(G.nodes())
+        self.Aadj = nx.adjacency_matrix(G, nodelist=self.node_list)
+
+        # We need a schema to check compatibility with GraphSAGE, GAT, GCN
```
Maybe remove reference to GraphSAGE? Can this generator be used with GraphSAGE?

youph replied:
Fixed.

youph replied:
fixed.

PantelisElinas commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+
+        Args:
+            layer_sizes: list of output sizes of GAT layers in the stack
+            attn_heads: number of attention heads
```
Maybe explain that this specifies the number of attention heads for all layers expect the last one that always has 1 head.

youph replied:
done, done.

PantelisElinas commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+        Args:
+            layer_sizes: list of output sizes of GAT layers in the stack
+            attn_heads: number of attention heads
+            attn_heads_reduction:
```
Missing description.

youph replied:
description added

adocherty commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+        """
+
+        Args:
+            layer_sizes: list of output sizes of GAT layers in the stack
```
It's a good idea to specify the accepted types of the arguments. For example (list of int) or (List[int]).

youph replied:
fixed

adocherty commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+            self.attn_heads_reduction = ["concat"] * (len(layer_sizes) - 1) + [
+                "average"
+            ]
+        else:
```
Perhaps a check that the length of `attn_heads_reduction` is correct would be a good idea.

youph replied:
check is added

adocherty commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+            ), "{}: generator must be of type FullBatchNodeGenerator or None; received object of type {} instead".format(
+                type(self).__name__, type(generator).__name__
+            )
+
```
I agree!

youph replied:
Good point! Thanks for the link to the discussion.

adocherty commented on tests/layer/test_graph_attention.py:
```diff
@@ -0,0 +1,375 @@
+        assert model.output_shape[-1] == self.F_out
+
+        X = np.ones((self.N, self.F_in))  # features
+        A = np.eye(self.N)  # adjacency matrix with self-loops only
```
Perhaps a test with a non-identity adjacency matrix would be useful?

youph replied:
what would that test check?

adocherty replied:
Check the application of the attention layer to a graph with some neighbours.

youph replied:
added one such (simple) test with nodes 0-1 connected

adocherty commented on tests/mapper/test_node_mappers.py:
```diff
@@ -559,3 +559,88 @@ def test_hinnodemapper_no_neighbors():
+
+        [X, A], y, sample_weights = gen.__getitem__(0)
+        assert np.array_equal(X, gen.features)  # X should be equal to gen.features
+        assert (A != gen.A).nnz == 0  # A should be equal to gen.A
```
`len(np.nonzero(A != gen.A)[0])` would work for both dense and sparse matrices :-)

youph replied:
fixed

adocherty commented on demos/node-classification-gat/gat-cora-example.py:
```diff
@@ -0,0 +1,360 @@
+    if args.interface == "fit":
+        all_predictions = model.predict(x=[X, A], batch_size=N)
+    else:
+        all_predictions = model.predict_generator(generator.flow(node_ids))
```
The `predict` or `predict_generator` function now operates differently to the `GraphSAGE` model in that if you give it less than the complete set of nodes, it will still return all predictions and in a fixed order.

This is confusing for users – whatever is given inthe `flow` argument will be ignored by the `predict_generator` as it's just used for masking.

youph replied:
I've added a comment about this in both the demos

youph commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+
+"""
+Definition of Graph Attention Network (GAT) layer and GAT class that is a stack of GAT layers
+"""
```
Reference added

youph commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+        kernel_constraint=None,
+        bias_constraint=None,
+        attn_kernel_constraint=None,
+        **kwargs
```
Added "Must be a Keras constraint https://keras.io/constraints/" to the `Args` docstring

youph commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+        # Adding self-loops to A via setting the diagonal elements of A to 1.0:
+        # N = kwargs.get("num_nodes")
+        # get the number of nodes from inputs[1] directly, rather than passing it via kwargs
+        N = inputs[1]._keras_shape[-1]
```
No, this doesn't work, as this gives a tensor `Tensor("graph_attention_1/strided_slice:0", shape=(), dtype=int32)` instead of an integer.

adocherty replied:
Ah, if you need the int shape you can use `K.int_shape(...)`

youph replied:
thanks!

youph commented on stellargraph/layer/graph_attention.py:
```diff
@@ -0,0 +1,481 @@
+        N = inputs[1]._keras_shape[-1]
+        if N is not None:
+            # create self-loops
+            A = tf.linalg.set_diag(A, K.cast(np.ones((N,)), dtype="float"))
```
I've added an argument "add_self_loops" to `GAT.node_model` method, which toggles adding self-loops in the `GraphAttention` layers of the GAT model. The default value is set to `True` so that the user does not have to worry about adding self-loops to make sure the resulting GAT layers match those in the paper, yet still allowing advanced users to switch self-loops off if needed.

youph commented:
> * The predictions obtained using `predict_generator` ignore the arguments to the `.flow` method of the generator. This could be confusing for users, and is different from the behaviour of the GraphSAGE model.
I've added a comment to both demos, explaining this difference.

> * The number of attention heads at the final layer is set to 1 always.
Fixed: now it can be either a single int (in which case it's that number for all but the last layer, and one for the last layer), or a list of ints.

> * I like the model checkpointing and early stopping in the python script, but I think it's overkill for the notebook – it slows down the execution of the notebook and also prints out a lot of epochs when running it.
True, although I wanted to have few more new things in the notebook demo, compared to those using GraphSAGE for the same task. Otherwise the difference boils down to 2 lines of code: different generator and different model. Kind of boring. This way people who only look at notebooks (and not the scripts) would see how to organise early stopping and checkpointing :)

> * Also note that the model is different from the published GAT model: they have the final layer output the number of classes and apply a softmax over that. We add a dense layer. The accuracy isn't much different between either option, but maybe we can point out that we are doing it differently.
I've changed the model architecture in the notebook, to match that of the GAT paper.

youph commented:
> Hi @youph
> 
> it looks good. Have a look at my comments.
> 
> I noticed that running the notebook does not produce the same result as the reference paper. I get about 80% accuracy and the paper states 83%. It looks like the parameter setting are the same. Maybe later we should have a look at why there is a difference.
> 
> Regards,
> 
> P.

@PantelisElinas that's true; one possible reason is a somewhat different model architecture. I've updated the model architecture in the notebook demo, to match that of the paper more closely.

adocherty commented:
Hi @youph .. The changes look good to me. 

In terms of documentation, we'll need to add the GAT class to the `docs/api.txt` file in a similar way to GraphSAGE or GCN.  To make the automatic API documentation work, all documentation for a class needs to be under the main class definition, not in the `__init__` method.

You can check the docs locally by doing the following:
```
pip install sphinx_rtd_theme sphinx
cd docs
make html
open _build/html/index.html
```

youph commented:
@adocherty thanks for the instructions about the docs; I've updated the docs api, and brushed the doc strings.

youph commented:
@wangzhen263 Please modify the doc strings in GCN layer and class as well, to make sure they are displayed nicely in the docs, as per @adocherty's instructions above. In particular, to make the automatic API documentation work, all documentation for a class needs to be under the main class definition, not in the `__init__` method.
